### ISSUE TITLE
MVI + retain view + screen rotation

### ISSUE THREAD

**Issue by MFlisar** [sockeqwe/mosby-conductor#19]
I'm not sure if this is desired, but imho, if the controller has `setRetainViewMode(RetainViewMode.RETAIN_DETACH)`, the last known state should not be reemitted again after screen rotation because the view was not destroyed but only detached and reattached and therefore, it has already rendered the last known state in this case

I know it's still in beta state and this may just not have been implemented yet, but even in this case, I wanted to document this behaviour

**Comment by sockeqwe** [comment]
Thanks for the hint.

> the view was not destroyed but only detached and reattached and therefore, it has already rendered the last known state in this case

I'm not sure if this is true because the presenter is detached from the view. But while detached, the presenter could still be subscribed to some background stuff that updates the state. 

I haven't done too much with `setRetainViewMode(RetainViewMode.RETAIN_DETACH)` but I think it only keeps the `View` so that you don't have to reinflate the xml layout again, but I'm not sure if you can update UI widgets while they are not attached to a Window / Activity. For example, I do have doubts that layouting would work (which would be triggered on each state change) in the case when the view is not attached to a window.

Anyway, if there would be some extra steps for `setRetainViewMode(RetainViewMode.RETAIN_DETACH)` I think it would make things internally a lot more complicated and error prone. Therefore, I think I would prefer to stick on the same behavior as before without RETAIN_DETACH. Do you see any performance issue if the state is rerendered again or have any other concerns?

**Comment by MFlisar** [comment]
The issue I have is flickering sometimes...  Normally, rendering the state does not cost much time (mostly), but updating a lot of images may result in short flickering. And this (that rerendering does not cost much) may not be true for everyone...

For my personal use case I have a solution:
For starting animations, my view already has a flag indicating "I'm ready conductor, you can start animating between the views". So I just use this state and ignore the view state, if this flag is already true and my data is not empty (which is only the case after rotation with an retained view instance).

*Imho, the correct solution would be following:*
* on rotation, pause emitting NEW states
* after rotation, don't reemit the last known state, instead, start emitting buffered states (only if there are any) that were emitted during the rotation...

This probably adds quite some complexity, I know, but it would be the correct solution, wouldn't it? At least, I would have an rx operator that is capable of pausing/resuming events we could use for that behaviour, that may be useable here (I used it for the same problem in my RXBus library to avoid emitting events while an activity/fragment is paused or a view is detached and similar...)

I'm not sure if it really is necessary, currently, simple ignoring the state after rotation works as well, you just have to manually find out if the state is the first state after rotation or not (which probably is most of the time much less complicated?). But it must be done manually by the user and imho, it's an unexpected behaviour...

**Comment by sockeqwe** [comment]
> The issue I have is flickering sometimes

This should be avoided with some change detections. i.e. it is known that `textView.setText("Foo")` will cause a complete re-rendering even if the text this view displays is already "Foo". so you better do something like `if (!textView.getText().equals("Foo")) textView.setText("Foo");`. You might have to do that for some other critical UI widgets too. Please note that this is not only MVI related, but perhabs "rendering" calls happen more often in MVI.

> but updating a lot of images may result in short flickering

Which image library do you use? Usually when you call "load that image" it is returned directly from memory cache, so there should no flickering at all. Just a random guess? Are you seeing this issue on screen orientation changes and is it because the ImageView has a different size now and you use a "resize" operator in your image loading library? If yes, then it is not really related to this issue ...

> Imho, the correct solution would be following:
> - on rotation, pause emitting NEW states
> - after rotation, don't reemit the last known state, instead, start emitting buffered states (only if there are any) that were emitted during the rotation...

`setRetainViewMode(RetainViewMode.RETAIN_DETACH)` doesn't work on screen orientation changes. It only keeps the view when you are navigating through the back stack. For example:
Controller A is currently displayed with `RetainViewMode.RETAIN_DETACH`. Pushing Controller B on top of the back stack will conductor force to keep the View of Controller A somewhere in memory. When pressing the back button Controller B is popped and Controller A appears again, this time without inflating the xml layout, but reusing the view previously kept in memory.

After a screen orientation change the View will **always** be inflated again from XML layout regardless of `RetainViewMode.RETAIN_DETACH`

So we are really just talking about the moment when Controller A is not "visible" because another controller is on top of the back stack. And here certainly Presenter A (of controller A) could receive a new ViewState  which then must be  re-rendered once the view is reattached.

**Comment by MFlisar** [comment]
OK, I thought the view is retained on orientation change as well... I think I got this now. And of course, updating a not visible view is just fine, as you say and nothing must be changed for that.

Then it seems like my current solution is anyway what you would do as well (detect if state was rendered already, if true, skip rendering it again) and that's it.

Still I see, that after rotating my device I do not have to render anything, but this would be the same without `setRetainViewMode(RetainViewMode.RETAIN_DETACH)`, correct? I have to check that, maybe only my adapter is not recreated so it still has it's list of `ViewHolders` and therefore does not need to be updated...

**Comment by sockeqwe** [comment]
> OK, I thought the view is retained on orientation change as well

This would create memory leaks because the original view has been created with previous Activity   (before the screen orientation change). Activity still gets destroyed (conductor is not doing something magic here) and an entirely new Activity instance will be created during screen orientation changes.


> Then it seems like my current solution is anyway what you would do as well (detect if state was rendered already, if true, skip rendering it again) and that's it.

If you just want to skip one rendering, this is as easy :

```java
class MyController extends MviController<MyView, MyViewState> {
     private MyViewState lastState;
     ...

    @Overrid
    public void render(MyViewState newState){
        if (newState.equals(lastState)
            return;
       
        // manipulate your UI widgets
        ...
        lastState = newState;
    }
}
```

> but this would be the same without setRetainViewMode(RetainViewMode.RETAIN_DETACH), correct?

Yes, because RetainViewMode.RETAIN_DETACH doesn't work / has no impact on screen orientation changes.

>  it still has it's list of ViewHolders and therefore does not need to be updated...

This would be a huge memory leak because the ViewHolders has been created with the activity (context / layoutinflater) of the previous activity (before the screen orientation change)



----
To sum it up. I think mosby-conductor is working as expected and therefore I close this issue.
